sql/src/main/java/io/crate/jobs/KeepAliveTimers.java

```diff
@@ -0,0 +1,132 @@
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.jobs;
+
+import io.crate.action.job.KeepAliveRequest;
+import io.crate.action.job.TransportKeepAliveAction;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.inject.Singleton;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportResponse;
+
+import java.util.UUID;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+@Singleton
+public class KeepAliveTimers {
+
+    private static final ESLogger LOGGER = Loggers.getLogger(KeepAliveTimers.class);
+
+    private final ThreadPool threadPool;
+    private final TimeValue maxKeepAliveTime;
+    private final TransportKeepAliveAction transportKeepAliveAction;
+
+    @Inject
+    public KeepAliveTimers(ThreadPool threadPool,
+                           @JobContextService.JobKeepAlive TimeValue maxKeepAliveTime,
+                           TransportKeepAliveAction transportKeepAliveAction) {
+        this.threadPool = threadPool;
+        this.maxKeepAliveTime = maxKeepAliveTime;
+        this.transportKeepAliveAction = transportKeepAliveAction;
+    }
+
+    public ResettableTimer forJobOnNode(final UUID jobId, final String nodeId) {
+        return forRunnable(new Runnable() {
+            @Override
+            public void run() {
+                transportKeepAliveAction.keepAlive(nodeId, new KeepAliveRequest(jobId), new ActionListener<TransportResponse.Empty>() {
+                    @Override
+                    public void onResponse(TransportResponse.Empty empty) {
+                        LOGGER.trace("keep alive for job [{}] on downstream [{}] finished successfully", jobId, nodeId);
+                    }
+
+                    @Override
+                    public void onFailure(Throwable e) {
+                        LOGGER.trace("keep alive for job [{}] on downstream [{}] failed", e,jobId, nodeId);
+                    }
+                });
+            }
+        });
+    }
+
+    public ResettableTimer forRunnable(Runnable runnable) {
+        TimeValue delay = TimeValue.timeValueMillis(Math.max(1, maxKeepAliveTime.millis() / 3));
+        return forRunnable(runnable, delay);
+    }
+
+    public ResettableTimer forRunnable(Runnable runnable, TimeValue delay) {
+        return new ResettableTimer(threadPool, runnable, delay);
+    }
+
+
+    public static class ResettableTimer {
+        private final Runnable runnable;
```

Comment:
can you rename the variables here to be a bit more specific? what is the runnable supposed to do?
What kind of future does the futureHolder hold?